src/backend/distributed/executor/intermediate_results.c

```diff
@@ -0,0 +1,699 @@
+	EState *estate = NULL;
+	List *nodeList = NIL;
+	bool writeLocalFile = true;
+	RemoteFileDestReceiver *resultDest = NULL;
+	ParamListInfo paramListInfo = NULL;
+
+	CheckCitusVersion(ERROR);
+
+	estate = CreateExecutorState();
+	resultDest = CreateRemoteFileDestReceiver(resultKeyString, estate, nodeList,
+											  writeLocalFile);
+
+	ExecuteQueryStringIntoDestReceiver(queryString, paramListInfo,
+									   (DestReceiver *) resultDest);
+
+	FreeExecutorState(estate);
+
+	PG_RETURN_INT64(resultDest->tuplesSent);
+}
+
+
+/*
+ * CreateRemoteFileDestReceiver creates a DestReceiver that streams results
+ * to a set of worker nodes.
+ */
+static RemoteFileDestReceiver *
+CreateRemoteFileDestReceiver(char *resultKey, EState *executorState,
+							 List *initialNodeList, bool writeLocalFile)
+{
+	RemoteFileDestReceiver *resultDest = NULL;
+
+	resultDest = (RemoteFileDestReceiver *) palloc0(sizeof(RemoteFileDestReceiver));
+
+	/* set up the DestReceiver function pointers */
+	resultDest->pub.receiveSlot = RemoteFileDestReceiverReceive;
+	resultDest->pub.rStartup = RemoteFileDestReceiverStartup;
+	resultDest->pub.rShutdown = RemoteFileDestReceiverShutdown;
+	resultDest->pub.rDestroy = RemoteFileDestReceiverDestroy;
+	resultDest->pub.mydest = DestCopyOut;
+
+	/* set up output parameters */
+	resultDest->resultKey = resultKey;
+	resultDest->executorState = executorState;
+	resultDest->initialNodeList = initialNodeList;
+	resultDest->memoryContext = CurrentMemoryContext;
+	resultDest->writeLocalFile = writeLocalFile;
+
+	return resultDest;
+}
+
+
+/*
+ * RemoteFileDestReceiverStartup implements the rStartup interface of
+ * RemoteFileDestReceiver. It opens the relation
+ */
+static void
+RemoteFileDestReceiverStartup(DestReceiver *dest, int operation,
+							  TupleDesc inputTupleDescriptor)
+{
+	RemoteFileDestReceiver *resultDest = (RemoteFileDestReceiver *) dest;
+
+	const char *resultKey = resultDest->resultKey;
+
+	CopyOutState copyOutState = NULL;
+	const char *delimiterCharacter = "\t";
+	const char *nullPrintCharacter = "\\N";
+
+	List *initialNodeList = resultDest->initialNodeList;
+	ListCell *initialNodeCell = NULL;
+	List *connectionList = NIL;
+	ListCell *connectionCell = NULL;
+
+	resultDest->tupleDescriptor = inputTupleDescriptor;
+
+	/* define how tuples will be serialised */
+	copyOutState = (CopyOutState) palloc0(sizeof(CopyOutStateData));
+	copyOutState->delim = (char *) delimiterCharacter;
+	copyOutState->null_print = (char *) nullPrintCharacter;
+	copyOutState->null_print_client = (char *) nullPrintCharacter;
+	copyOutState->binary = CanUseBinaryCopyFormat(inputTupleDescriptor);
+	copyOutState->fe_msgbuf = makeStringInfo();
+	copyOutState->rowcontext = GetPerTupleMemoryContext(resultDest->executorState);
+	resultDest->copyOutState = copyOutState;
+
+	resultDest->columnOutputFunctions = ColumnOutputFunctions(inputTupleDescriptor,
+															  copyOutState->binary);
+
+	BeginOrContinueCoordinatedTransaction();
```

Comment:
It would make sense to only do this if the worker list is `NIL`

You can actually write both locally and to workers, we're just not providing a UDF that does that. The reasoning is that if we support local execution then you could have a CTE that is referenced in both a distributed query and a local query.

Reply:
Actually, probably better to always start a coordinated transaction, since we may first have a local-only result and then a workers-only result, and we don't want to store them in different places.